### PR TITLE
SONARJAVA-1043 - Improve Rule S2259 Add support of NP_EQUALS_SHOULD_H…

### PR DESCRIPTION
**Pull request by DidierBesset**
…ANDLE_NULL_ARGUMENT

### CODE REVIEW

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Change this comparison to use the equals method.](https://github.com/SonarSource/sonar-java/blob/c1c388c38a9f6636ab2e03ff4477ae5003b685bd/java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java#L250) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)

File: java-squid/src/main/java/org/sonar/java/se/checks/NullDereferenceCheck.java [hunk 1]
```diff
@@ -118,20 +115,8 @@ private static boolean isAnnotatedCheckForNull(MethodInvocationTree syntaxNode)
...
-    }
-    if (expressionTree != null) {
-      if (expressionTree.is(Tree.Kind.IDENTIFIER)) {
-        name = ((IdentifierTree) expressionTree).name();
-      } else if (expressionTree.is(Tree.Kind.METHOD_INVOCATION)) {
-        name = ((MethodInvocationTree) expressionTree).symbol().name();
-      }
-    }
-    return name;
+    SyntaxTreeNameFinder finder = new SyntaxTreeNameFinder();
```
Comment (benzonico) [inline java-squid/src/main/java/org/sonar/java/se/checks/NullDereferenceCheck.java]: 0o How is that linked to the ticket you are solving ?

File: java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java [hunk 2]
```diff
@@ -185,8 +187,7 @@ public ProgramState apply(ProgramState input) {
         }
       });
 
-      if (variableTree.symbol().metadata().isAnnotatedWith("javax.annotation.CheckForNull")
-        || variableTree.symbol().metadata().isAnnotatedWith("javax.annotation.Nullable")) {
+      if (isEqualsMethod || parameterCanBeNull(variableTree)) {
```
Comment (benzonico) [inline java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java]: not usually found of downcasting but in this case it is better than redefining the way to detect equals method: You can use : https://github.com/SonarSource/sonar-java/blob/master/java-squid/src/main/java/org/sonar/java/model/declaration/MethodTreeImpl.java#L326-326 by downcasting to MethodTreeImpl in this case : 
`((MethodTreeImpl) tree).isEqualsMethod`

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Change this comparison to use the equals method.](https://github.com/SonarSource/sonar-java/blob/82b7c314652583ee13f7efc87aff069af1b9e42b/java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java#L250) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)

**Comment by DidierBesset** [comment]
How to ask for re-submition of the travis-PR job because I think the error is not at all related to this PR?

**Comment by DidierBesset** [comment]
Scan enabled, but I am just afraid it will eat up more cpu...

I want to use this class in my other checker, so keep it outside.
Scan enbaled, but I am just afraid it will eat up more cpu...

I want to use this class in my other checker, so keep it outside.

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Change this comparison to use the equals method.](https://github.com/SonarSource/sonar-java/blob/e94bd9291242cc89d6b5a21eea9a7debd197620f/java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java#L250) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Change this comparison to use the equals method.](https://github.com/SonarSource/sonar-java/blob/8926bdf4cd7f70c3c68f7f127e19057b0477b303/java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java#L268) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Change this comparison to use the equals method.](https://github.com/SonarSource/sonar-java/blob/d7855e335713a7a38bd5e00162ee2154a0d375bf/java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java#L268) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)

File: java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java [hunk 3]
```diff
@@ -212,6 +213,11 @@ public ProgramState apply(ProgramState input) {
     return startingStates;
   }
 
+  private static boolean parameterCanBeNull(final VariableTree variableTree) {
+    return variableTree.symbol().metadata().isAnnotatedWith("javax.annotation.CheckForNull")
```
Comment (mpaladin) [inline java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java]: @DidierBesset you can assign `variableTree.symbol().metadata()` to a variable and simplify return statement.

File: java-squid/src/test/files/se/NullDereferenceCheck.java [hunk 4]
```diff
@@ -603,4 +603,9 @@ void while_loop() {
       currentParent = null;
     }
   }
+
+  @Override
+  public boolean equals(Object obj) {
+    return getValue() == ((MyClass) obj).getValue(); // Noncompliant {{NullPointerException might be thrown as 'obj' is nullable here}}
+  }
 }
```
Comment (mpaladin) [inline java-squid/src/test/files/se/NullDereferenceCheck.java]: @DidierBesset EOF new line.

File: java-squid/src/main/java/org/sonar/java/se/checks/NullDereferenceCheck.java (hunk 1 above)
Comment (mpaladin) [inline java-squid/src/main/java/org/sonar/java/se/checks/NullDereferenceCheck.java]: @DidierBesset you could move this method as a static method in `SyntaxTreeNameFinder` so you will be able to use it with a single call.

**Pull request merged**